Remove BitmapImage::updateFromSettings()

shallawa opened the pull request:
#### 0c6288be7ffe1f75fcd39a6bd2aa10ee6afa993f
<pre>
Remove BitmapImage::updateFromSettings()
<a href="https://bugs.webkit.org/show_bug.cgi?id=267607">https://bugs.webkit.org/show_bug.cgi?id=267607</a>
<a href="https://rdar.apple.com/121077174">rdar://121077174</a>

Reviewed by Cameron McCormack.

Calculate all the image painting options before calling BitmapImage::draw().
Remove all the calls to BitmapImage::updateFromSettings() and pass the calculated
AllowImageSubsampling and ShowDebugBackground through ImagePaintingOptions.

The first frame of an animated image had to be decoded synchronously. In fact
this restriction is not needed. The rules to decode the static image and the
first frame of an animated image should be the same.

Late decoded frames of an animated images was showing a yellow rectangle if
showDebugBackground is enabled. This is not really needed since the animation
does not advance until the next frame is fully decoded.

* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::drawImage):
* Source/WebCore/platform/graphics/BitmapImage.cpp:
(WebCore::BitmapImage::draw):
(WebCore::BitmapImage::shouldUseAsyncDecodingForAnimatedImages const):
(WebCore::BitmapImage::subsamplingLevelForScaleFactor):
(WebCore::BitmapImage::internalStartAnimation):
(WebCore::BitmapImage::advanceAnimation):
(WebCore::BitmapImage::decode):
(WebCore::BitmapImage::updateFromSettings): Deleted.
* Source/WebCore/platform/graphics/BitmapImage.h:
* Source/WebCore/platform/graphics/ImagePaintingOptions.h:
(WebCore::ImagePaintingOptions::allowImageSubsampling const):
(WebCore::ImagePaintingOptions::showDebugBackground const):
(WebCore::ImagePaintingOptions::setOption):
* Source/WebCore/platform/graphics/ImageTypes.h:
* Source/WebCore/rendering/BackgroundPainter.cpp:
(WebCore::BackgroundPainter::paintFillLayer):
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
(WebCore::RenderBoxModelObject::decodingModeForImageDraw const):
* Source/WebCore/rendering/RenderImage.cpp:
(WebCore::RenderImage::paintIntoRect):
* Source/WebCore/rendering/svg/RenderSVGImage.cpp:
(WebCore::RenderSVGImage::paintIntoRect):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::imageFrameIndex):

Canonical link: <a href="https://commits.webkit.org/273528@main">https://commits.webkit.org/273528@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8f3509fe8133d7933dae4ad6f8096f7fc9041339

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/35659 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/14600 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/37797 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/38396 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/32119 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/36857 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/16988 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/11626 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/30906 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/36212 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/12340 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/31732 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/10845 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/10847 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/31873 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/39641 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/32403 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/32209 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/36804 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/11038 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/8929 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/34885 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/12757 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8144 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/11546 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/11825 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebCore/platform/graphics/BitmapImage.cpp:
```diff
@@ -393,7 +385,7 @@ bool BitmapImage::canUseAsyncDecodingForLargeImages() const
 
 bool BitmapImage::shouldUseAsyncDecodingForAnimatedImages() const
 {
-    return canAnimate() && m_allowAnimatedImageAsyncDecoding && (shouldUseAsyncDecodingForTesting() || m_source->canUseAsyncDecoding());
```
Can you explain why we no longer need `m_allowAnimatedImageAsyncDecoding` and no longer consult `shouldUseAsyncDecodingForTesting()`?

shallawa replied:
Because they are consulted in `RenderBoxModelObject::decodingModeForImageDraw()`.  This makes the DecodingMode for the animated images and the large images are done the same way,

heycam commented on Source/WebCore/platform/graphics/BitmapImage.cpp:
```diff
@@ -247,12 +239,12 @@ ImageDrawResult BitmapImage::draw(GraphicsContext& context, const FloatRect& des
     RefPtr<NativeImage> image;
-    if (options.decodingMode() == DecodingMode::Asynchronous) {
-        ASSERT(!m_currentFrame || !canAnimate());
+    if (!canAnimate() && options.decodingMode() == DecodingMode::Asynchronous) {
```
Can you explain this change, is it a change of behavior?

shallawa replied:
We used to pass DecodingMode::Synchronous for animated images always in `RenderBoxModelObject::decodingModeForImageDraw()`

```
    auto* bitmapImage = dynamicDowncast<BitmapImage>(image);
    if (!bitmapImage || bitmapImage->canAnimate()) {
        // The DecodingMode for the current frame has to be Synchronous. The DecodingMode
        // for the next frame will be calculated in BitmapImage::internalStartAnimation().
        return DecodingMode::Synchronous;
    }
```

This is to avoid going through the then part of this if-statement. The else part of this if- statement was handling the async for animated images through `BitmapImage::internalStartAnimation()`

```
    if (shouldUseAsyncDecodingForAnimatedImages()) {
        ...
        m_source->requestFrameAsyncDecodingAtIndex
   }
```

This change fix this unclear behavior by making `RenderBoxModelObject::decodingModeForImageDraw()` passes the correct DecodingMode from the beginning. So this is not a behavior change.

heycam commented on Source/WebCore/platform/graphics/BitmapImage.cpp:
```diff
@@ -521,12 +517,8 @@ void BitmapImage::advanceAnimation()
     if (!frameIsBeingDecodedAndIsCompatibleWithOptionsAtIndex(nextFrame, DecodingOptions(DecodingMode::Asynchronous)))
         internalAdvanceAnimation();
-    else {
-        // Force repaint if showDebugBackground() is on.
-        if (m_showDebugBackground)
-            imageObserver()->changedInRect(*this);
```
Can you explain why we no longer need this?

shallawa replied:
The animated image wants to advance to the next frame. But we do not advance to the next frame util it is fully decoded. In this case, the decoding takes too much time. The timer fire but the frame is not ready yet. So we have to wait till the decoding is complete and once we receive `imageFrameAvailableAtIndex()` from the work queue, we draw this frame.

The deleted code was showing a yellow rectangle in this case even though the current frame is rendered. So do we really need to show the yellow rectangle in this case? I do not think so. But I can put back.

heycam replied:
OK, I agree we don't want the yellow rectangle in this case.

heycam replied:
Since this is a behavior change, I think it is worth mentioning in the commit message.

heycam commented on Source/WebCore/rendering/RenderBoxModelObject.cpp:
```diff
@@ -305,23 +305,21 @@ bool RenderBoxModelObject::hasAutoHeightOrContainingBlockWithAutoHeight() const
         return DecodingMode::Synchronous;
-    }
 
     // Some document types force synchronous decoding.
```
This comment should move up with the `document().isImageDocument()` check.

shallawa replied:
I do not quite understand your review here. I think we can even remove this comment since it does not add much.

heycam commented on Source/WebCore/rendering/RenderBoxModelObject.cpp:
```diff
@@ -360,6 +358,18 @@ DecodingMode RenderBoxModelObject::decodingModeForImageDraw(const Image& image,
+        if (bitmapImage->currentFrameIndex())
+            return DecodingMode::Asynchronous;
+
+        return defaultDecodingMode();
```
To clarify, this is a behavior change for the first frame, is that right? Whereas before, we would always sync decode the first frame of an animated image, but now we will check e.g. if it's visible in the viewport to determine whether to sync decode?

If that's right, please mention this (and any other intended behavior changes) in the commit message.

shallawa replied:
Yes it is a behavior change. And I think it is the only behavior change in this PR. But I think it's the correct behavior. I think the old behavior was done because it was easier to do it this way. There should not be a reason for disallowing us from async decoding the first frame of an animated image.

I will add a comment about this in the commit message.

shallawa committed:
Remove BitmapImage::updateFromSettings()
https://bugs.webkit.org/show_bug.cgi?id=267607
rdar://121077174

Reviewed by Cameron McCormack.

Calculate all the image painting options before calling BitmapImage::draw().
Remove all the calls to BitmapImage::updateFromSettings() and pass the calculated
AllowImageSubsampling and ShowDebugBackground through ImagePaintingOptions.

The first frame of an animated image had to be decoded synchronously. In fact
this restriction is not needed. The rules to decode the static image and the
first frame of an animated image should be the same.

Late decoded frames of an animated images was showing a yellow rectangle if
showDebugBackground is enabled. This is not really needed since the animation
does not advance until the next frame is fully decoded.

* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::drawImage):
* Source/WebCore/platform/graphics/BitmapImage.cpp:
(WebCore::BitmapImage::draw):
(WebCore::BitmapImage::shouldUseAsyncDecodingForAnimatedImages const):
(WebCore::BitmapImage::subsamplingLevelForScaleFactor):
(WebCore::BitmapImage::internalStartAnimation):
(WebCore::BitmapImage::advanceAnimation):
(WebCore::BitmapImage::decode):
(WebCore::BitmapImage::updateFromSettings): Deleted.
* Source/WebCore/platform/graphics/BitmapImage.h:
* Source/WebCore/platform/graphics/ImagePaintingOptions.h:
(WebCore::ImagePaintingOptions::allowImageSubsampling const):
(WebCore::ImagePaintingOptions::showDebugBackground const):
(WebCore::ImagePaintingOptions::setOption):
* Source/WebCore/platform/graphics/ImageTypes.h:
* Source/WebCore/rendering/BackgroundPainter.cpp:
(WebCore::BackgroundPainter::paintFillLayer):
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
(WebCore::RenderBoxModelObject::decodingModeForImageDraw const):
* Source/WebCore/rendering/RenderImage.cpp:
(WebCore::RenderImage::paintIntoRect):
* Source/WebCore/rendering/svg/RenderSVGImage.cpp:
(WebCore::RenderSVGImage::paintIntoRect):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::imageFrameIndex):

Canonical link: https://commits.webkit.org/273528@main
